cupy/testing/helper.py

```diff
@@ -352,15 +352,47 @@ def _convert_output_to_ndarray(c_out, n_out, sp_name, check_sparse_format):
             type(c_out), type(n_out)))
 
 
+def _resolve_tolerance(type_check, result, rtol, atol):
+    def _resolve(dtype, tol):
+        if isinstance(tol, dict):
+            tol1 = tol.get(dtype.type)
```

Comment:
If `tol` is specified with a wrong format, for example `{'float16': 1e-3}`, the default tolerance will be chosen and may pass the wrong test unexpectedly. Did you have some ideas to avoid this case?

Reply:
Definitely. I think an option is using an association list `[(<dtype>, <tolerance>)]` instead of a dictionary, where `<dtype>` is an object such that `numpy.dtype(<dtype>) == {numpy,cupy}_result.dtype` is satisfied. This way we can use `[(numpy.float16, 1e-3)]`, `[('float16', 1e-3)]` or `[('f', 1e-3)]` as well. How do you think?

Reply:
Hmm, I feel the `dict` is a more natural interface. How about checking if all keys are of `type` type or `default` during the preprocess?

Reply:
Sounds good. I'll fix the code as such.

Reply:
I've made a commit, PTAL!